strawberryfields/io.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Blackbird I/O
+=============
+
+This module contains functions for loading and saving Strawberry
+Fields :class:`~.Program` objects from/to Blackbird scripts.
+
+Summary
+-------
+
+.. autosummary::
+    to_blackbird
+    to_program
+    save
+    load
+
+Code details
+~~~~~~~~~~~~
+"""
+# pylint: disable=protected-access,too-many-nested-blocks
+import os
+import blackbird
+
+from . import ops
+from .program import Program
+
+
+def to_blackbird(prog, version="1.0"):
+    """Convert a Strawberry Fields Program to a Blackbird Program.
+
+    Args:
+        prog (Program): the Strawberry Fields program
+        version (str): Blackbird script version number
+
+    Returns:
+        blackbird.BlackbirdProgram:
+    """
+    bb = blackbird.BlackbirdProgram(name=prog.name, version=version)
+
+    if prog.backend is not None:
+        # set the target
+        bb._target["name"] = prog.backend
+
+    # fill in the quantum circuit
+    for cmd in prog.circuit:
+        op = {"kwargs": {}, "args": []}
+
+        op["op"] = cmd.op.__class__.__name__
+        op["modes"] = [i.ind for i in cmd.reg]
+
+        if "Measure" in op["op"]:
+            # special case to take into account 'select' keyword argument
+            if cmd.op.select is not None:
+                op["kwargs"]["select"] = cmd.op.select
+
+            if cmd.op.p:
+                # argument is quadrature phase
+                op["kwargs"]["phi"] = cmd.op.p[0].x
+
+        else:
+            if cmd.op.p is not None:
+                for a in cmd.op.p:
+                    # check if reg ref transform
+                    if isinstance(a.x, ops.RegRefTransform):
+                        if a.x.func_str is not None:
+                            # Note: not currently guaranteed to work,
```

Comment:
What isn't guaranteed to work? Can you give a failure case?

Reply:
Because the `func_str` is set by the user, and doesn't need to correlate to the RegRefTransform function at all :laughing:.

For example, `Dgate(RR(q[0], lambda x: x**2, func_str='hi'))`.

This is really something that can only be solved by allowing symbolic functions, i.e.,
```python
q0 = sympy.Symbol('q0')
Dgate(q0**2) | q[1]
```

We can then access a string representation of the function,

```python
>>> func = q0**2`
>>> str(func)
q0**2
```

or use the Lambdify ability of SymPy to turn it into a evaluate-able Python function:

```python
>>> f = sympy.lambdify(q[0], func)
>>> f(5)
25
```